File: jar/src/main/java/jline/lib/butools/dmap/RandomDMMAP.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 * Ported from BUTools V2.0
 */
package jline.lib.butools.dmap;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import jline.lib.butools.mc.DTMCSolve;
import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class RandomDMMAP {
    private RandomDMMAP() {}

    /**
     * Returns a random discrete marked Markovian arrival process.
     */
    public static MatrixCell randomDMMAP(int order, int types, double mean, int zeroEntries,
                                         int maxTrials, double prec, Random random) {
        if (types < 1) {
            throw new IllegalArgumentException("RandomDMMAP: 'types' must be positive integer!");
        }

        if (zeroEntries > (order + 1) * (order - 1) + types * (order * order - 1)) {
            throw new IllegalArgumentException("RandomDMMAP: Too many zero entries requested!");
        }

        // Generate all possible zero distributions among rows
        List<int[]> zeroDistr = allZeroDistributions(order, zeroEntries);

        int trials = 0;
        while (trials < maxTrials) {
            List<Integer> indices = new ArrayList<Integer>();
            for (int i = 0; i < zeroDistr.size(); i++) indices.add(i);
            Collections.shuffle(indices, random);

            for (int zdix : indices) {
                int[] zDistr = zeroDistr.get(zdix);

                boolean bad = false;
                for (int z : zDistr) {
                    if (z >= (types + 1) * order - 1) {
                        bad = true;
                        break;
                    }
                }
                if (bad) {
                    trials++;
                    continue;
                }

                Matrix B = new Matrix(order, (types + 1) * order);
                for (int i = 0; i < order; i++) {
                    List<Integer> rp = new ArrayList<Integer>();
                    for (int j = 0; j < (types + 1) * order - 1; j++) rp.add(j);
                    Collections.shuffle(rp, random);
                    double[] a = new double[(types + 1) * order - 1];
                    for (int j = 0; j < (types + 1) * order - 1 - zDistr[i]; j++) {
                        a[rp.get(j)] = random.nextDouble();
                    }
                    for (int j = 0; j < i; j++) {
                        B.set(i, j, a[j]);
                    }
                    for (int j = i + 1; j < (types + 1) * order; j++) {
                        B.set(i, j, a[j - 1]);
                    }
                }

                MatrixCell D = new MatrixCell(types + 1);
                double[] sc = new double[order];
                for (int kk = 0; kk <= types; kk++) {
                    D.set(kk, new Matrix(order, order));
                    for (int i = 0; i < order; i++) {
                        for (int j = 0; j < order; j++) {
                            D.get(kk).set(i, j, B.get(i, kk * order + j));
                            sc[i] += D.get(kk).get(i, j);
                        }
                    }
                }

                boolean anyZero = false;
                for (double v : sc) {
                    if (v == 0.0) {
                        anyZero = true;
                        break;
                    }
                }
                if (anyZero) continue;

                for (int kk = 0; kk <= types; kk++) {
                    for (int i = 0; i < order; i++) {
                        for (int j = 0; j < order; j++) {
                            D.get(kk).set(i, j, D.get(kk).get(i, j) / sc[i]);
                        }
                    }
                }

                Matrix sumD = new Matrix(order, order);
                for (int kk = 0; kk <= types; kk++) {
                    sumD = sumD.add(D.get(kk));
                }

                Matrix I = Matrix.eye(order);
                if (D.get(0).rank() == order && I.sub(sumD).rank() == order - 1) {
                    Matrix alpha = DTMCSolve.dtmcSolve(sumD);
                    double minAlpha = Double.POSITIVE_INFINITY;
                    for (int j = 0; j < order; j++) {
                        minAlpha = Math.min(minAlpha, Math.abs(alpha.get(0, j)));
                    }

                    if (minAlpha > prec) {
                        boolean fullZero = false;
                        for (int kk = 0; kk <= types; kk++) {
                            boolean allZero = true;
                            outer:
                            for (int i = 0; i < order; i++) {
                                for (int j = 0; j < order; j++) {
                                    if (D.get(kk).get(i, j) != 0.0) {
                                        allZero = false;
                                        break outer;
                                    }
                                }
                            }
                            if (allZero) {
                                fullZero = true;
                                break;
                            }
                        }

                        if (!fullZero) {
                            double[] d = new double[order];
                            for (int idx = 0; idx < order; idx++) d[idx] = random.nextDouble();

                            MatrixCell Dv = new MatrixCell(types + 1);
                            for (int kk = 0; kk <= types; kk++) {
                                Dv.set(kk, new Matrix(order, order));
                                for (int i = 0; i < order; i++) {
                                    for (int j = 0; j < order; j++) {
                                        Dv.get(kk).set(i, j, (1.0 - d[i]) * D.get(kk).get(i, j));
                                    }
                                }
                            }
                            for (int i = 0; i < order; i++) {
                                Dv.get(0).set(i, i, Dv.get(0).get(i, i) + d[i]);
                            }

                            double[] m = MarginalMomentsFromDMMAP.marginalMomentsFromDMMAP(Dv, 1, prec);

                            for (int i = 0; i < order; i++) {
                                d[i] = 1.0 - (1.0 - d[i]) * m[0] / mean;
                            }

                            for (int kk = 0; kk <= types; kk++) {
                                for (int i = 0; i < order; i++) {
                                    for (int j = 0; j < order; j++) {
                                        D.get(kk).set(i, j, (1.0 - d[i]) * D.get(kk).get(i, j));
                                    }
                                }
                            }
                            for (int i = 0; i < order; i++) {
                                D.get(0).set(i, i, D.get(0).get(i, i) + d[i]);
                            }

                            if (CheckDMMAPRepresentation.checkDMMAPRepresentation(D, prec)) {
                                return D;
                            }
                        }
                    }
                }
                trials++;
            }
        }

        throw new IllegalStateException("No feasible random DMMAP found!");
    }

    public static MatrixCell randomDMMAP(int order, int types, double mean, int zeroEntries,
                                         int maxTrials, double prec) {
        return randomDMMAP(order, types, mean, zeroEntries, maxTrials, prec, new Random());
    }

    public static MatrixCell randomDMMAP(int order, int types, double mean, int zeroEntries, int maxTrials) {
        return randomDMMAP(order, types, mean, zeroEntries, maxTrials, 1e-7, new Random());
    }

    public static MatrixCell randomDMMAP(int order, int types, double mean, int zeroEntries) {
        return randomDMMAP(order, types, mean, zeroEntries, 1000, 1e-7, new Random());
    }

    public static MatrixCell randomDMMAP(int order, int types, double mean) {
        return randomDMMAP(order, types, mean, 0, 1000, 1e-7, new Random());
    }

    public static MatrixCell randomDMMAP(int order, int types) {
        return randomDMMAP(order, types, 10.0, 0, 1000, 1e-7, new Random());
    }

    private static List<int[]> allZeroDistributions(int states, int zeros) {
        if (states == 1) {
            return Collections.singletonList(new int[]{zeros});
        }
        List<int[]> result = new ArrayList<int[]>();
        for (int iz = 0; iz <= zeros; iz++) {
            List<int[]> subDistributions = allZeroDistributions(states - 1, zeros - iz);
            for (int[] sub : subDistributions) {
                int[] combined = new int[sub.length + 1];
                System.arraycopy(sub, 0, combined, 0, sub.length);
                combined[sub.length] = iz;
                java.util.Arrays.sort(combined);
                boolean exists = false;
                for (int[] e : result) {
                    if (java.util.Arrays.equals(e, combined)) {
                        exists = true;
                        break;
                    }
                }
                if (!exists) {
                    result.add(combined);
                }
            }
        }
        return result;
    }
}

File: jar/src/main/kotlin/jline/lib/butools/dmap/RandomDMMAP.kt
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 * Ported from BUTools V2.0
 */
package jline.lib.butools.dmap

import jline.lib.butools.mc.dtmcSolve
import jline.util.matrix.Matrix
import jline.util.matrix.MatrixCell
import java.util.Random
import kotlin.math.abs

/**
 * Returns a random discrete marked Markovian arrival process.
 *
 * @param order The size of the DMMAP
 * @param types The number of different arrival types
 * @param mean The mean inter-arrival time of the DMMAP (default: 10.0)
 * @param zeroEntries The number of zero entries in the D matrices (default: 0)
 * @param maxTrials Maximum number of trials to find a proper DMMAP (default: 1000)
 * @param prec Numerical precision for checking irreducibility (default: 1e-7)
 * @param random Random number generator
 * @return The D0...Dtypes matrices of the DMMAP
 */
fun randomDMMAP(
    order: Int,
    types: Int,
    mean: Double = 10.0,
    zeroEntries: Int = 0,
    maxTrials: Int = 1000,
    prec: Double = 1e-7,
    random: Random = Random()
): MatrixCell {
    if (types < 1) {
        throw IllegalArgumentException("RandomDMMAP: 'types' must be positive integer!")
    }

    if (zeroEntries > (order + 1) * (order - 1) + types * (order * order - 1)) {
        throw IllegalArgumentException("RandomDMMAP: Too many zero entries requested!")
    }

    // Generate all possible zero distributions among rows
    val zeroDistr = allZeroDistributions(order, zeroEntries)

    var trials = 0
    while (trials < maxTrials) {
        val indices = (0 until zeroDistr.size).shuffled(random)

        for (zdix in indices) {
            val zDistr = zeroDistr[zdix]

            var bad = false
            for (z in zDistr) {
                if (z >= (types + 1) * order - 1) {
                    bad = true
                    break
                }
            }
            if (bad) {
                trials++
                continue
            }

            val B = Matrix(order, (types + 1) * order)
            for (i in 0 until order) {
                val rp = (0 until (types + 1) * order - 1).shuffled(random)
                val a = DoubleArray((types + 1) * order - 1)
                for (j in 0 until (types + 1) * order - 1 - zDistr[i]) {
                    a[rp[j]] = random.nextDouble()
                }
                for (j in 0 until i) {
                    B[i, j] = a[j]
                }
                for (j in i + 1 until (types + 1) * order) {
                    B[i, j] = a[j - 1]
                }
            }

            val D = MatrixCell(types + 1)
            val sc = DoubleArray(order)
            for (k in 0..types) {
                D[k] = Matrix(order, order)
                for (i in 0 until order) {
                    for (j in 0 until order) {
                        D[k][i, j] = B[i, k * order + j]
                        sc[i] += D[k][i, j]
                    }
                }
            }

            if (sc.any { it == 0.0 }) continue

            for (k in 0..types) {
                for (i in 0 until order) {
                    for (j in 0 until order) {
                        D[k][i, j] = D[k][i, j] / sc[i]
                    }
                }
            }

            var sumD = Matrix(order, order)
            for (k in 0..types) {
                sumD = sumD.add(D[k])
            }

            val I = Matrix.eye(order)
            if (D[0].rank() == order && I.sub(sumD).rank() == order - 1) {
                val alpha = dtmcSolve(sumD)
                var minAlpha = Double.POSITIVE_INFINITY
                for (j in 0 until order) {
                    minAlpha = minOf(minAlpha, abs(alpha[0, j]))
                }

                if (minAlpha > prec) {
                    var fullZero = false
                    for (k in 0..types) {
                        var allZero = true
                        outer@ for (i in 0 until order) {
                            for (j in 0 until order) {
                                if (D[k][i, j] != 0.0) {
                                    allZero = false
                                    break@outer
                                }
                            }
                        }
                        if (allZero) {
                            fullZero = true
                            break
                        }
                    }

                    if (!fullZero) {
                        val d = DoubleArray(order) { random.nextDouble() }

                        val Dv = MatrixCell(types + 1)
                        for (k in 0..types) {
                            Dv[k] = Matrix(order, order)
                            for (i in 0 until order) {
                                for (j in 0 until order) {
                                    Dv[k][i, j] = (1.0 - d[i]) * D[k][i, j]
                                }
                            }
                        }
                        for (i in 0 until order) {
                            Dv[0][i, i] += d[i]
                        }

                        val m = marginalMomentsFromDMMAP(Dv, 1, prec)

                        for (i in 0 until order) {
                            d[i] = 1.0 - (1.0 - d[i]) * m[0] / mean
                        }

                        for (k in 0..types) {
                            for (i in 0 until order) {
                                for (j in 0 until order) {
                                    D[k][i, j] = (1.0 - d[i]) * D[k][i, j]
                                }
                            }
                        }
                        for (i in 0 until order) {
                            D[0][i, i] += d[i]
                        }

                        if (checkDMMAPRepresentation(D, prec)) {
                            return D
                        }
                    }
                }
            }
            trials++
        }
    }

    throw IllegalStateException("No feasible random DMMAP found!")
}

private fun allZeroDistributions(states: Int, zeros: Int): List<IntArray> {
    if (states == 1) {
        return listOf(intArrayOf(zeros))
    }
    val result = mutableListOf<IntArray>()
    for (iz in 0..zeros) {
        val subDistributions = allZeroDistributions(states - 1, zeros - iz)
        for (sub in subDistributions) {
            val combined = (sub.toList() + iz).sorted().toIntArray()
            if (result.none { it.contentEquals(combined) }) {
                result.add(combined)
            }
        }
    }
    return result
}
